### ISSUE TITLE
[Discussion]  Approaches to HTML table generation

### ISSUE THREAD

**Issue by nyurik** [vega/vega#1258]
This ticket is to gather requirements and solicit ideas on how to create a new html-table oriented renderer, possibly as a standalone Vega plugin.  Once finalized, it will probably be moved to a new repo.

## Goal
Allow users to generate HTML tables from the data after being processed by the Vega transformations, including interactivity.

## Approaches
One possible approach is to have a new `cell` mark. Having this mark automatically creates an HTML table. The table could be positioned either underneath the <canvas> or <svg> element, or on top of it _(TBD - multiple layers?)_.  Multiple cell marks could target either one or multiple output tables. Using groups it may be possible to draw tables inside other tables.

## Sizing
Table should occupy the entire space of the container (autosize=fit), or be the same size as the graph (autosize=none and pad).

## Parameters
* `name` -- ID of the mark
* `table` -- ID of the table - multiple marks could have the same table ID
* `column` and `row`  (instead of x,y) - integer cell position within a table.  The table will dynamically grow whenever the column or row values exceed currently allocated space.
* `rowspan` and `colspan` -- same as HTML table's params.
* From the [text mark](https://vega.github.io/vega/docs/marks/text/): `align`, `baseline`, `dir`, `ellipsis`, `font`, `fontSize`, `fontWeight`, `fontStyle`, `limit`, `text`
`border` -- TBD

## Open questions
* Table-level parameters - should there be a top-level and group-level `tables` section, similar to data? Table nesting based on groups?
* Multiple tables vs a single table - should there always be a single table (similar to how width/height always refer to the main drawing area), or should it be possible to draw multiple ones, and allow table positioning?
* specifying headers/footers - could be done by a separate `cell` mark targeting the same table and using bold font

**Comment by domoritz** [comment]
I love this. One concern I have is that text wrapping doesn't work in Vega charts right now. Just something to keep in mind.

**Comment by g3o2** [comment]
But if it was rendered in html there would be text wrapping, so there would be no need for canvas or svg, where the said issue arises, would there?

**Comment by iliatimofeev** [comment]
I see two different use cases:
### Simple table in a single chart like annotations. 
In this case will be great to allow link it somehow with legend. It is quite useful for business data presentation. Excel chart example:
![chart_excel_data](https://user-images.githubusercontent.com/11561581/39897576-7acb07cc-54bb-11e8-8a4c-ccb6e17ddf65.png)


### Complex dashboard where table in interactive and reflects to the actions in other parts of dashboard. 

For dashboard case after our discussion about tile maps support I came to strange idea: create another layer above vega, vega-light to describe a dashboard as set of interconnected components. Each of them could be vega-light or vega chart or Leaflet based map or interactive table. So we can be less restricted with how exactly component is rendered to html and focus on event management interface and data management. This approach seems quite extendable using wrappers over existing components. I saw some technical solutions on interconnection between different vega chat on mail list.  So it seems possible to implement.


Maybe it is off topic here, but I think that approach would simplify requirements to native table in Vega.

**Comment by nyurik** [comment]
@iliatimofeev hi, I think your example (table as an axis) could be done with the current Vega without any new functionality. In it, you use Vega's built-in alignment mechanism (axis with band scaling) to position things.  Aligning an HTML table with Vega content seems to be too error-prone to be worth it (it could be done in theory, but the moment things start resizing, or user zooms-in, it could go crazy).

A plugin could grab the output of a specific data source, and using some configuration settings creates an HTML table. I am not exactly sure if it would be possible to get user events back into Vega for this one, but perhaps it could be done using selectors?

If this is done with a new type of a mark as described at the top, it would be easier to get the events, but may require substantially more dev efforts.

**Comment by nyurik** [comment]
@iliatimofeev all of my use cases are the same -- users can only access Vega spec itself, not JS or HTML.  It is my job as a developer to allow them to create needed visualizations without sacrificing the security of the system.  A Vega plugin or a `cell` mark are two ways to achieve that.

The fundamental difference between vega & html is the layout engine. In html, you rely on the browser to position things. In Vega, you decide the exact location of each mark, or rely on a relatively simple built-in Vega positioning logic for axis/legends.  HTML tables do all the coplex magic like deciding the size of the cell depending on the cell content, wrapping, relative column widths, etc.  In theory, it is possible to force HTML table to have very stringent sizes (e.g. set pixel perfect width/height of each cell), but then there is really not much point of using HTML as oppose to directly drawing things with Vega, e.g. [this example](https://vega.github.io/vega-lite/examples/layer_text_heatmap.html).

The real benefit of HTML mostly comes from the layout capabilities and the ability to select and copy text. The layout implies that you let browser do all the heavy lifting, but that also means many things won't be able to align in a meaningful way.

**Comment by tonne1** [comment]
@iliatimofeev: I fully subscribe to the use case you outlined above under "Complex dashboard". I have built a report/dashbord generator based on d3 and Firebird, with reusable interactive and interacting components (including treemaps, boxplots, sankey - and tables) and of course user interaction. Now I am investigating if this can be done using vega/ preferrably vega-lite, i.e., on a more abstract, general level.

From my experience tables are the most versatile and complex visual components, and any visualisation engine which does not provide for tables is only half useful. Some users prefer graphics while others find it easier to relate to text and numbers (dimensions and measures). Tables can arrange heterogenous content which belongs together (row context/column context). After all, we are mainly dealing with tabular data. Apart from that, tables can have these properties:

- row / row group and column aggregates (subtotals and totals)
- column groups (multi-line header)
- hierarchical levels of data, which can be expanded/collapsed (levels of detail)
- sorting by column headers, including sorting by more than one column, in order
- hiding columns
- selecting rows as filter for other visual components
- individual cells as starting point to drill into other reports, show/hide other components
- table cells can host other visual components like sparklines, bars, heatmap bars, trend arrows ...
- they can be the target of user interaction with other components (e.g. line chart with time brush)
- complex styling
- they can be huge
- ...

Re building an engine like vega this means a design decision like you have indicated: You either go the full way and generate all of the above or you create an interface capable of controlling these things in an external structure and delegate some of the tasks. Or both :-)

A big issue with using tables to arrange (= layout) other visual components is dependency. While nesting components can be very useful and sometimes unavoidable, the price is that they have to be created in order - and request their data in this order, too.

This is not an easy one.

**Comment by domoritz** [comment]
An idea from @jakevdp and me was to build tables as an external module. The main issue I see with tables in Vega is support for text wrapping and flexible table layouts. HTML already offers all of the things we need. The external module could use HTML to create a table for a Vega data source. When the data source changes, the table would update automatically. The Vega view API provides access to this feature.

**Comment by g3o2** [comment]
Text wrapping seems to be a big thing in tables indeed. But is it _really_ that important or is it just a limited use case? vega-tooltip already comes in handy for showing the full value and more when needed.

So why not just do it without html? Why not leverage the existing vega technology instead? This would allow for a few quite unique features by combining cell marks with other marks through composition and interaction.

Think e.g. of brushing; clicking on cell values to remove/add other marks; highlighting another mark by increasing its size or color when hovering over a table cell; draw callouts which link a point mark to a (couple of) table cell(s) and vice versa, ...

If you only need stand-alone tables, as a sort of annotation or web Excel, then I'd  tend towards @domoritz 's approach. Maybe there could be even a roadmap for both approaches.

**Comment by nyurik** [comment]
@g3o2 it's not just wrapping - it's the layout itself that's the hard part.  HTML can resize columns based on the widest value, and has a lot of code dedicated to various edge cases.  In Vega, you can only draw strings at a given position. But you need to calculate that position - that's the really hard part.  If you think in your usecase you know the position already, you don't need any new features - you can already draw the table with the current VL & V.

I agree that having it as part of layout is a much better option from composition perspective. It would be great to draw text as a table with sparklines. And that should be the ultimate goal. But that requires Vega to create multiple SVG or canvases together with HTML - a fairly complex undertaking given the existing state of Vega technology.

**Comment by domoritz** [comment]
I see my suggestion an interim solution that's quick to implement but will work well. It won't support all the features we might want eventually but it gives us a testbed to see what we really need.

**Comment by domoritz** [comment]
I'm not sure whether to add this to embed or not but yes, like that.

**Comment by bphenriques** [comment]
I have similar requirements as described here in this issue. I want to to design tabular views in Kibana through Vega and export that same information to CSV.

**Comment by SDxKeeper** [comment]
Can it be interactive to the level of pivottable? Where user can drag and drop column names to dynamically update transform?

**Comment by djouallah** [comment]
Any update on this,  I start using vega lite via altair,  and I love this library and used it already in production, even as a non programmer I managed to create a cross filter dashboard which is a statement to the easy of the API,  now with all the complex charts available, table and pivot are still the most used visuals to show financial data,  it will be great if vega can support that,  just simple table with multiple metrics,  and cross filter.

**Comment by farhanchaudhry** [comment]
This would be a great feature. waiting for it.

**Comment by brianzar** [comment]
This would be an extremely useful (and awesome) feature to add! Is it on the roadmap at all?

**Comment by domoritz** [comment]
Not right now but we would love your contributions. However, support for connecting to Vega-Lite selections is on our roadmap: https://docs.google.com/document/d/1fscSxSJtfkd1m027r1ONCc7O8RdZp1oGABwca2pgV_E/edit#.

**Comment by tonyxiao** [comment]
Is it possible to access the transformed data to render table outside of vega?

**Comment by nyurik** [comment]
@tonyxiao yes https://vega.github.io/vega/docs/api/debugging/#inspect-data-sets
